### PR TITLE
refactor: update Angular to compile cleanly with TypeScript 2.6

### PR DESCRIPTION
**Pull request by chuckjaz**
## PR Type
What kind of change does this PR introduce?

```
[x] Refactoring (no functional changes, no api changes)
```

## What is the current behavior?

Angular an only use 2.5 of TypeScript to compile.

## What is the new behavior?

Angular can use either 2.5 or 2.6 of TypeScript to compile everything except the language service which needs to be updated with the upgrad as the language service API has breaking changes.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/common/src/i18n/locale_en.ts [hunk 1]
```diff
@@ -9,6 +9,13 @@
 // THIS CODE IS GENERATED - DO NOT MODIFY
```
Comment (mprobst) [inline packages/common/src/i18n/locale_en.ts]: Any idea on this? I was also totally puzzled by the failure mode of not removing the type arguments to this function, I'm not even sure what bit of our toolchain got that wrong.

File: packages/compiler-cli/src/metadata/bundle_index_host.ts [hunk 2]
```diff
@@ -44,7 +44,7 @@ function createSyntheticIndexHost<H extends ts.CompilerHost>(
 
   newHost.writeFile =
       (fileName: string, data: string, writeByteOrderMark: boolean,
-       onError?: (message: string) => void, sourceFiles?: ts.SourceFile[]) => {
+       onError: (message: string) => void|undefined, sourceFiles: ReadonlyArray<ts.SourceFile>) => {
```
Comment (mprobst) [inline packages/compiler-cli/src/metadata/bundle_index_host.ts]: does `onError` now return `void|undefined`? Or is this a typo/unintended change?

File: packages/compiler-cli/src/metadata/schema.ts [hunk 3]
```diff
@@ -99,7 +99,11 @@ export function isFunctionMetadata(value: any): value is FunctionMetadata {
 }
 
 export type MetadataValue = string | number | boolean | undefined | null | MetadataObject |
-    MetadataArray | MetadataSymbolicExpression | MetadataError;
+    MetadataArray | MetadataSymbolicExpression | MetadataSymbolicReferenceExpression |
```
Comment (mprobst) [inline packages/compiler-cli/src/metadata/schema.ts]: When I made this change, I wasn't sure whether `MetadataSymbolicExpression` should maybe just be the union of all the symbol expression types.

File: packages/compiler-cli/src/perform_watch.ts [hunk 4]
```diff
@@ -161,7 +161,7 @@ export function performWatchCompilation(host: PerformWatchHost):
       const originalWriteFileCallback = cachedCompilerHost.writeFile;
       cachedCompilerHost.writeFile = function(
           fileName: string, data: string, writeByteOrderMark: boolean,
-          onError?: (message: string) => void, sourceFiles?: ts.SourceFile[]) {
+          onError: (message: string) => void|undefined, sourceFiles: ReadonlyArray<ts.SourceFile>) {
```
Comment (mprobst) [inline packages/compiler-cli/src/perform_watch.ts]: see comment above for the odd `|undefined`

File: packages/compiler-cli/src/transformers/compiler_host.ts [hunk 5]
```diff
@@ -266,9 +266,9 @@ export class TsCompilerAotCompilerTypeCheckHostAdapter implements ts.CompilerHos
     }
     // Note: This can also return undefined,
```
Comment (mprobst) [inline packages/compiler-cli/src/transformers/compiler_host.ts]: remove the outdated comment here?

**Review by mprobst — APPROVED** [review]

**Comment by chuckjaz** [comment]
Merged into #21144

**Pull request closed without merging**